pkg/daemon/update.go

```diff
@@ -1726,6 +1731,57 @@ func (dn *Daemon) updateSSHKeys(newUsers []ign3types.PasswdUser) error {
 		}
 	}
 	if !dn.mock {
+		// On RHCOS8 and in older versions of OKD, the keys are written to `/home/core/.ssh/authorized_keys`.
+		// Newer versions of OKD will expect the keys at `/home/core/.ssh/authorized_keys.d/ignition`.
+		// Remove the keys file from the old path here. It will be recreated at the correct path by the
+		// atomicallyWriteSSHKey function called right after.
+		authKeyPath := filepath.Join(coreUserSSHPath, "authorized_keys")
+		authKeyFragmentDirPath := filepath.Join(coreUserSSHPath, "authorized_keys.d")
+
+		if dn.os.IsFCOS() {
+			// Check if the authorized_keys file exists. If it does, remove it.
+			_, err := os.Stat(authKeyPath)
+			if err == nil {
+				err := os.RemoveAll(authKeyPath)
+				if err != nil {
+					return fmt.Errorf("failed to remove path '%s': %v", authKeyPath, err)
+				}
+			} else if !os.IsNotExist(err) {
+				// This shouldn't ever happen
+				return fmt.Errorf("unexpectedly failed to get info for path '%s': %v", authKeyPath, err)
+			}
+
+			// Ensure authorized_keys.d/ignition is the only fragment that exists
+			keyFragmentsDir, err := ioutil.ReadDir(authKeyFragmentDirPath)
+			if err == nil {
+				for _, fragment := range keyFragmentsDir {
+					if fragment.Name() != "ignition" {
+						keyPath := filepath.Join(authKeyFragmentDirPath, fragment.Name())
+						err := os.RemoveAll(keyPath)
+						if err != nil {
+							return fmt.Errorf("failed to remove path '%s': %v", keyPath, err)
+						}
+					}
+				}
+			} else if !os.IsNotExist(err) {
+				// This shouldn't ever happen
+				return fmt.Errorf("unexpectedly failed to get info for path '%s': %v", authKeyFragmentDirPath, err)
+			}
+		} else {
```

Comment:
Since we are not handling RHCOS 9, we can skip else section.

Reply:
umm, or this could be needed like you said during downgrade but we haven't started thinking about this move yet. I haven't tested myself but do you know if it can cause undesired behavior if both /home/core/.ssh/authorized_keys.d/ignition and /home/core/.ssh/authorized_keys are present on a node?